Updates to DataFrame.__new__ causes issue with custom DataFrame Generic types

**Describe the bug**

Hi all, first off thanks for this effort! We use `pandas-stubs` in the [`pandera`](https://pandera.readthedocs.io/en/stable/) project for optional mypy type linting and it's been working great!

Recent changes ([here](https://github.com/pandas-dev/pandas-stubs/commit/147fbe6a060da0c1acf28ce530e2d39518039f0f)) have caused issues in pandera type annotation, where we use a pandera `SchemaModel` in a subclassed DataFrame class like so:

```python
class MySchema(pandera.SchemaModel): ...

# then you can initialize a validated dataframe like so:
vaildated_data = pandera.typing.DataFrame[MySchema](data)
```

However, this will raise a mypy error if `pandas-stubs` is installed:
```
Type application has too few types (2 expected)
```

**To Reproduce**

Consider this code, which subclasses dataframes :
```python
# bar.py
from typing import Generic, TypeVar

import pandas as pd

T = TypeVar("T")

class MyDataFrame(pd.DataFrame, Generic[T]):
    ...


def func() -> MyDataFrame[int]:
    return MyDataFrame[int]({"foo": [1, 2, 3]})
```

Then using `mypy`:
```
mypy bar.py

# bar.py:14: error: Type application has too few types (2 expected)  [misc]
# Found 1 error in 1 file (checked 1 source file)
```

I've isolated the cause of this [here](https://github.com/pandas-dev/pandas-stubs/commit/147fbe6a060da0c1acf28ce530e2d39518039f0f#diff-2a2c7ce81de1d491f11217e36077830686177277dd96884fead786f22b222bf1R166-R170), where removing the `ListLike` type annotations (including the one in the `Iterable` type) makes the error go away. I don't understand why the type annotations in the `__new__` method would effect the type of the subclassed generic arguments 🤷‍♂️

**Please complete the following information:**
 - OS: MacOS
 - OS Version: 10.15.7
 - python version: 3.9.12
 - version of type checker: mypy==0.971
 - version of installed `pandas-stubs`: 1.4.3.220815


**Additional context**

This issue doesn't show up in the previous version `pandas-stubs==1.4.3.220807`

--- a/pandas-stubs/_typing.pyi
+++ b/pandas-stubs/_typing.pyi
@@ -99,6 +99,7 @@ AxisType = Literal["columns", "index", 0, 1]
 DtypeNp = TypeVar("DtypeNp", bound=np.dtype[np.generic])
 KeysArgType = Any
 ListLike = TypeVar("ListLike", Sequence, np.ndarray, "Series", "Index")
+ListLikeU = Union[Sequence, np.ndarray, Series, Index]
 StrLike = Union[str, np.str_]
 Scalar = Union[
     str,

--- a/pandas-stubs/core/frame.pyi
+++ b/pandas-stubs/core/frame.pyi
@@ -61,6 +61,7 @@ from pandas._typing import (
     Label,
     Level,
     ListLike,
+    ListLikeU,
     MaskType,
     Renamer,
     Scalar,
@@ -157,10 +158,10 @@ class DataFrame(NDFrame, OpsMixin):
 
     def __new__(
         cls,
-        data: ListLike
+        data: ListLikeU
         | DataFrame
         | dict[Any, Any]
-        | Iterable[tuple[Hashable, ListLike]]
+        | Iterable[tuple[Hashable, ListLikeU]]
         | None = ...,
         index: Axes | None = ...,
         columns: Axes | None = ...,

--- a/tests/test_frame.py
+++ b/tests/test_frame.py
@@ -8,10 +8,12 @@ from typing import (
     TYPE_CHECKING,
     Any,
     Callable,
+    Generic,
     Hashable,
     Iterable,
     Iterator,
     Tuple,
+    TypeVar,
     Union,
 )
 
@@ -1593,3 +1595,16 @@ def test_dict_items() -> None:
     # GH 180
     x = {"a": [1]}
     check(assert_type(pd.DataFrame(x.items()), pd.DataFrame), pd.DataFrame)
+
+
+def test_generic() -> None:
+    # GH 197
+    T = TypeVar("T")
+
+    class MyDataFrame(pd.DataFrame, Generic[T]):
+        ...
+
+    def func() -> MyDataFrame[int]:
+        return MyDataFrame[int]({"foo": [1, 2, 3]})
+
+    func()

